src/library/coverartdelegate.cpp

```diff
@@ -108,16 +107,16 @@ void CoverArtDelegate::paintItem(QPainter *painter,
     info.hash = index.sibling(index.row(), m_iCoverHashColumn).data().toUInt();
     info.trackLocation = index.sibling(index.row(), m_iTrackLocationColumn).data().toString();
 
+    double scaleFactor = getDevicePixelRatioF(dynamic_cast<QWidget*>(parent()));
     // We listen for updates via slotCoverFound above and signal to
     // BaseSqlTableModel when a row's cover is ready.
     QPixmap pixmap = pCache->requestCover(info, this, option.rect.width(),
-                                          m_bOnlyCachedCover, true);
+                                          scaleFactor, m_bOnlyCachedCover, true);
     if (!pixmap.isNull()) {
-        int width = math_min(pixmap.width(), option.rect.width());
-        int height = math_min(pixmap.height(), option.rect.height());
-        QRect target(option.rect.x(), option.rect.y(),
-                     width, height);
-        QRect source(0, 0, target.width(), target.height());
+        int width = math_min(pixmap.width(), option.rect.width()) * scaleFactor;
```

Comment:
The `pixmap` dimensions are measured in device coords, but what about `option.rect`, `source`, `target`?

I agree with Daniel that this code seems to be incorrect, because it mixes up different measurements. The final multiplication scales the pixmap's coords up again.

Reply:
After some experimentation I realized `QPainter::drawPixmap(QRect, QPixmap, QRect)` wasn't working how I thought at all and it was also overcomplicated. The simpler `QPainter::drawPixmap(QPoint, QPixmap)` works fine without any complicated size conversions.